[api-minor] Render high-res partial page views when falling back to CSS zoom (bug 1492303)

nicolo-ribaudo opened the pull request:
Commit:

> **Render high-res partial page views when falling back to CSS zoom**
>
> When rendering big PDF pages at high zoom levels, we currently fall back to CSS zoom to avoid rendering canvases with too many pixels. This causes zoomed in PDF to look blurry, and the text to be potentially unreadable.
>
> This commit adds support for rendering _part_ of a page (called `PDFPageDetailView` in the code), so that we can render portion of a page in a smaller canvas without hiting the maximun canvas size limit.
>
> Specifically, we render an area of that page that is slightly larger than the area that is visible on the screen (100% larger in each direction, unless we have to limit it due to the maximum canvas size). As the user scrolls around the page, we re-render a new area centered around what is currently visible.

To play around with this patch, you need to use a large PDF (the one I'm using is https://www.gtt.to.it/cms/risorse/urbana/mappa/mapparete.pdf) and zoom in: in the current release the text becomes blurry, while with this patch it remains sharp. Also, if you set `maxCanvasPixels` to a smaller value (for example, 8M) you'll more clearly see the effects also at lower zoom levels. Note however that if `maxCanvasPixels` is set to less than 9 time the number of pixels visible in your window you'll get more frequent re-renders as you scroll.

The `PDFPageDetailView` class still process every single operation of the PDF, even if many of them will actually be _outside_ of the canvas and thus not be actually rendered. This is acceptable because we are only rendering one "detail view" per page at a time, and we are not splitting the page in multiple tiles that might appear at the same time on the screen. In the future however https://github.com/mozilla/pdf.js/pull/19043 _could_ improve the performance of this patch, and it would enable (for example) tiling a page into multiple detail views when printing it.

There are some very minor TODOs that I left in the code, for some areas I need review feedback for. Also, I still need to find the best way to write tests.

<details>
<summary>Fixes #14147</summary>

Before: 
<img width="989" alt="image" src="https://github.com/user-attachments/assets/3e0b45df-00d5-4e83-b30f-1b9571a65f09">

After: 
<img width="989" alt="image" src="https://github.com/user-attachments/assets/4d4e30f5-9fb8-4967-a0c8-21501c7cdc36">

</details>

<details>
<summary>https://www.gtt.to.it/cms/risorse/urbana/mappa/mapparete.pdf at 400% zoom</summary>

Before: 
<img width="989" alt="image" src="https://github.com/user-attachments/assets/61825ba3-e40a-43b2-8bff-1fd344658226">

After: 
<img width="989" alt="image" src="https://github.com/user-attachments/assets/6b501b52-2621-4b37-b76e-80f583cc3aae">


</details>

Fixes #14193

nicolo-ribaudo commented on web/pdf_page_view.js:
```diff
@@ -548,6 +700,8 @@ class PDFPageView {
 
     const childNodes = div.childNodes,
       zoomLayerNode = (keepZoomLayer && this.zoomLayer) || null,
+      detailScrollLayerNode =
+        (false && keepZoomLayer && this.detailView?.zoomLayer) || null,
```
Whops, I should remove this. Also, we should probably enable some lint rule against `false &&` and `true ||`.

nicolo-ribaudo commented:
After discussing this patch with @calixteman, I updated it to prioritize rendering the full css-zoomed canvas rather than the high-resolution one. This means that, when zooming in, you'll first see the css-zoomed canvas, and _then_ it will be replaced by the high-res one once it's ready. It's a slightly worse experience, however it guarantees that we do not regress in cases where the user starts moving around before that we are done rendering the background canvas (because if only the high-res canvas is there, they'll see just white until when the new high-res canvas is rendered).

I also changed the logic that decides when to re-render the high-res canvas to not only do it once the user scrolls past its edges, but when the user is _close to_ scrolling past the edges.

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -52,6 +52,85 @@ import { TextHighlighter } from "./text_highlighter.js";
 import { TextLayerBuilder } from "./text_layer_builder.js";
 import { XfaLayerBuilder } from "./xfa_layer_builder.js";
 
+// TODO: Should we make this an app option, or just always
+// enable the feature?
+const ENABLE_ZOOM_DETAIL = true;
```
Making this an `AppOption` seems like a good idea to me, since that may help future testing/debugging, however it probably does *not* need it to be a preference.
(In any case, leaving it as a code constant seems "wrong".)

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -52,6 +52,85 @@ import { TextHighlighter } from "./text_highlighter.js";
+// enable the feature?
+const ENABLE_ZOOM_DETAIL = true;
+
+class Recorder {
```
Is the intention to remove this class once the PR is ready to land, since the patch is already large and complex enough without this code?

Not to mention that there's a bunch logging that I don't believe is desirable in e.g. production-builds.

nicolo-ribaudo replied:
Yes, I plan to remove the second commit :) It's currently here because it has been very helpful when playing around with the PR, but it's definitely not something for production.

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -104,42 +183,275 @@ const DEFAULT_LAYER_PROPERTIES =
+  ["xfaLayer", 4],
 ]);
 
+class PDFPageViewBase {
```
Given the size and scope of this patch, it feels quite difficult to review as-is and catching any possible mistakes is going to be really difficult.

Hence I'll "insist" that this PR should have *separate commits* for easier reviewing:
 - A first commit that only introduces the `PDFPageViewBase` class.
 - A second commit that adds the actual "partial rendering" functionality.

nicolo-ribaudo replied:
Sounds good, I'll split it 👍

nicolo-ribaudo committed:
Extract `PDFPageViewBase` class out of `PDFPageView`

This base class contains the generic logic for:
- Creating a canvas and showing when appropriate
- Rendering in the canvas
- Keeping track of the rendering state

nicolo-ribaudo committed:
[api-minor] Render high-res partial page views when falling back to CSS zoom (bug 1492303)

When rendering big PDF pages at high zoom levels, we currently fall back
to CSS zoom to avoid rendering canvases with too many pixels. This
causes zoomed in PDF to look blurry, and the text to be potentially
unreadable.

This commit adds support for rendering _part_ of a page (called
`PDFPageDetailView` in the code), so that we can render portion of a
page in a smaller canvas without hiting the maximun canvas size limit.

Specifically, we render an area of that page that is slightly larger
than the area that is visible on the screen (100% larger in each
direction, unless we have to limit it due to the maximum canvas size).
As the user scrolls around the page, we re-render a new area centered
around what is currently visible.

nicolo-ribaudo commented on web/app.js:
```diff
@@ -2280,18 +2280,18 @@ if (typeof PDFJSDev === "undefined" || PDFJSDev.test("GENERIC")) {
   // If the page is (the most) visible when it starts rendering,
   // ensure that the page number input loading indicator is displayed.
-  if (pageNumber === this.page) {
+  if (pageNumber === this.page && !source.detailView) {
```
Note to myself: I need to change this, as the logic was for when the detail canvas was rendered _before_ the background canvas.

nicolo-ribaudo commented:
Unfortunately this approach has a problem right now: since the various SVG paths for drawings are inserted inside the `.canvasWrapper`, they are now _behind_ the detail layer and thus not visible. Marking as draft until this is fixed.

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -104,42 +179,255 @@ const DEFAULT_LAYER_PROPERTIES =
+  resume = null;
+
+  constructor(options) {
+    this.#enableHWA =
+      #enableHWA in options ? options.#enableHWA : options.enableHWA || false;
```
Can this be simplified e.g. as follows?
```suggestion
    this.#enableHWA = options.#enableHWA ?? (options.enableHWA || false);
```

nicolo-ribaudo replied:
Unfortunately no: if `#enableHWA` is not there (in the `super()` call from the `PDFPageView` constructor), `options.#enableHWA` will throw.

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -173,20 +464,15 @@ class PDFPageView {
     this.#annotationMode =
       options.annotationMode ?? AnnotationMode.ENABLE_FORMS;
     this.imageResourcesPath = options.imageResourcesPath || "";
+    this.enableDetailCanvas = AppOptions.get("enableDetailCanvas");
```
Sorry, but this isn't the "correct" way to handle options (note that the `maxCanvasPixels` case below is special for compatibility reasons).

Hence, please instead add the option in https://github.com/mozilla/pdf.js/blob/fbac4ab5f35dd4556069b69af51f78dbaaa62229/web/app.js#L471-L503 the same way that all existing `PDFViewer`/`PDFPageView` options are handled.

nicolo-ribaudo commented:
Fixing the drawing problem required changing the new canvas to be in the existing `.canvasWrapper`, rather than in a separate layer. This actually ended up simplifying the code a bit, and the `canvasWrapper` management logic is now fully contained in `PDFPageView`.

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -148,16 +334,19 @@ class PDFPageView {
-  #layers = [null, null, null, null];
+  #originalViewport = null;
+
+  #layers = [null, null, null, null, null];
```
Is this change actually needed now, given that the next patch was updated to place the detailView in the existing `canvasWrapper`?

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -110,36 +110,222 @@ const LAYERS_ORDER = new Map([
+    return this.#renderError;
+  }
+
+  _createCanvas(injectCanvas) {
```
To make it clearer that the parameter is a callback function, how about `_createCanvas(onShow) {` instead?

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -175,25 +364,20 @@ class PDFPageView {
 
     this._annotationCanvasMap = null;
 
+    this.canvasWrapper = null;
```
Given that the next patch was updated, can't we keep the private `this.#canvasWrapper` field now?

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -76,6 +76,10 @@ import { XfaLayerBuilder } from "./xfa_layer_builder.js";
+ * @property {boolean} [enableDetailCanvas] - When enabled, if the rendered
+ *   pages would need a canvas that is larger than `maxCanvasPixels`, it will
+ *   draw a second canvas on top of the CSS-zoomed one, that only renders the
+ *   part of the page that is close to the viewport.
```
Similar to all other parameters here, the comment should mention the default value.
```
The default value is `true`.
```

Snuffleupagus commented on web/pdf_viewer.js:
```diff
@@ -118,6 +118,10 @@ function isValidAnnotationEditorMode(mode) {
+ * @property {boolean} [enableDetailCanvas] - When enabled, if the rendered
+ *   pages would need a canvas that is larger than `maxCanvasPixels`, it will
+ *   draw a second canvas on top of the CSS-zoomed one, that only renders the
+ *   part of the page that is close to the viewport.
```
Similar to all other parameters here, the comment should mention the default value.
```
The default value is `true`.
```

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -362,6 +366,7 @@ class PDFPageView extends PDFPageViewBase {
     this.#annotationMode =
       options.annotationMode ?? AnnotationMode.ENABLE_FORMS;
     this.imageResourcesPath = options.imageResourcesPath || "";
+    this.enableDetailCanvas = options.enableDetailCanvas || false;
```
The default value should agree with `AppOptions`.
`this.enableDetailCanvas = options.enableDetailCanvas !== false;`

Snuffleupagus commented on web/pdf_rendering_queue.js:
```diff
@@ -119,12 +120,21 @@ class PDFRenderingQueue {
     if (numVisible === 0) {
       return null;
     }
+
```
Nit: This change looks unrelated.

Snuffleupagus commented on web/pdf_viewer.js:
```diff
@@ -312,6 +316,7 @@ class PDFViewer {
       this.removePageBorders = options.removePageBorders || false;
     }
     this.maxCanvasPixels = options.maxCanvasPixels;
+    this.enableDetailCanvas = options.enableDetailCanvas;
```
The default value should agree with `AppOptions`.
`this.enableDetailCanvas = options.enableDetailCanvas !== false;`

Snuffleupagus commented on web/pdf_viewer.js:
```diff
@@ -1226,7 +1232,13 @@ class PDFViewer {
       return;
     }
-    this.update();
+
+    const visible = this._getVisiblePages();
+    for (const { view, visibleArea } of visible.views) {
+      view.updateVisibleArea(visibleArea);
+    }
+
+    this.update(visible);
```
Maybe I'm not understanding the code, but why place the `updateVisibleArea`-call here rather than just placing all of this new code in the `update`-method?

Note that `update` is invoked in other spots as well, e.g. when changing scroll/spread modes (which may change the visible page area), and I'd assume that we always want to call `updateVisibleArea`!?

nicolo-ribaudo replied:
I didn't realize that the other callers of `.update()` can actually cause a change in the visible area. I was focusing on `.refresh()`, where _probably_ the area is going to remain the same.

Snuffleupagus reviewed (commented):
Leaving a few more comments, mostly for the *first* patch since I've only begun looking at the `Render high-res partial page views when falling back to CSS zoom` commit (and that one requires careful checking).

Also, is this new functionality anything that we could "easily" test with e.g. integration-tests?

nicolo-ribaudo commented on web/pdf_page_view.js:
```diff
@@ -148,16 +334,19 @@ class PDFPageView {
   };
 
-  #layers = [null, null, null, null];
+  #originalViewport = null;
```
Rebasing mistake :)

Snuffleupagus commented on web/pdf_rendering_queue.js:
```diff
@@ -119,12 +120,21 @@ class PDFRenderingQueue {
       return null;
     }
+
     for (let i = 0; i < numVisible; i++) {
       const view = visibleViews[i].view;
       if (!this.isViewFinished(view)) {
         return view;
       }
     }
+
+    for (let i = 0; i < numVisible; i++) {
+      const { detailView } = visibleViews[i].view;
+      if (detailView && !this.isViewFinished(detailView)) {
+        return detailView;
+      }
+    }
+
```
*Another observation/question that I have here is about the order of rendering.*

Would it make sense to attempt to render both a page *and* its `detailView` before moving onto the other visible pages, such that the most visible page will always render "fully" without having to wait for (potentially) a lot of other pages to finish rendering first?

Basically something like the following, where we combine the loops.
However, please note that I'm *not* saying that we must do this but I rather just want us to have considered/discussed this!
```suggestion
    for (let i = 0; i < numVisible; i++) {
      const view = visibleViews[i].view;
      if (!this.isViewFinished(view)) {
        return view;
      }
      const { detailView } = view;
      if (detailView && !this.isViewFinished(detailView)) {
        return detailView;
      }
    }

```

nicolo-ribaudo replied:
The answer is _maybe_.

The main use case I am thinking about for this patch is maps, because they are large documents where you usually want to zoom in a lot to the particular point that you are interested in.

The assumption is that the normal usage flow is:
- you zoom in a lot to roughly the point you are interested in
- you scroll a bit to adjust what's visible, so that you can actually see what you want
- you might scroll more in any direction, for example to follow a street or to see what is close to the point you were looking at

It's thus important that when you scroll around the page there is _something_ visible, and so this patch prioritizes rendering the "background view" over the "detail view" (a second reason is that it makes it easier to integrate https://github.com/mozilla/pdf.js/pull/19043 because it means that when we try to render the detail view we already have the bounding boxes, but there are also other ways to do it). This however matters _within a page_: it's probably unlikely that you zoom in to the wrong page, and then you need to scroll to another one. So your suggestion would be reasonable and not in conflict with the use case.

However, there is a problem: the visible pages are rendered first-to-last, meaning that if you have 10px of page 1 visible and 500px of page 2, it will first render page 1 and then page 2. Rendering page 1, then the detail of page 1, and then page 2 increase the total time for seeing _something meaningful_ on screen. So we would first need to change `getHighestPriority` to prioritize pages by how much they are visible, and _then_ we can do something like that you are suggesting.

> wait for (potentially) a lot of other pages

Note that it's _very likely_ that there are at most two pages on screen when the detail view is relevant: the pages must be so large that are above `maxCanvasPixels`, which makes it difficult to fit more than part of a page in the top half of the screen and part of it in the bottom half.

The only realistic case I can think of is when there are two pages side-by-side, so that we have four pages in total (one per "corner").

Snuffleupagus replied:
> So we would first need to change `getHighestPriority` to prioritize pages by how much they are visible, and _then_ we can do something like that you are suggesting.

Isn't that already the case though, given the `sortByVisibility` parameter?
https://github.com/mozilla/pdf.js/blob/2df5d8f3ef65a16ef0eb13668e64d74891b419d0/web/ui_utils.js#L423-L425

Snuffleupagus replied:
*As mentioned above, my intention here is only that we make an informed decision (whatever it may be) by considering the various cases at hand.*

> I have a slight preference for the second, because some PDFs can take very long (30-60 seconds per page) and so the quicker we show something the better it is, but ultimately I don't feel strongly about it.

There's only additional complication if there's multiple pages visible *and* they contain very large image resources.

We have API caching of operatorLists and image (and other) resources, however for complex pages those caches will be cleared (with a delay) after rendering completes. So by delaying rendering of the `detailView` until other pages have finished there's a risk that the caches have been cleared, and that we'll need to re-parse the page on the worker-thread for the `detailView` rendering.

See e.g. the following API code (and also the `_maybeCleanupAfterRender` property):
https://github.com/mozilla/pdf.js/blob/2df5d8f3ef65a16ef0eb13668e64d74891b419d0/src/display/api.js#L1780-L1787

nicolo-ribaudo commented:
I'm working on adding some meaningful tests.

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -1240,4 +1277,229 @@ class PDFPageView extends PDFPageViewBase {
+    this.pageView = pageView;
+    this.renderingId = "detail" + this.id;
+
+    this.detailArea = null;
```
Nit: Unless something "outside" of the class depends on this, can it be a private field instead?

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -1240,4 +1277,229 @@ class PDFPageView extends PDFPageViewBase {
+      // If there is already the lower resolution canvas behind,
+      // we don't show the new one until when it's fully ready.
+      this.pageView.renderingState === RenderingStates.FINISHED ||
+        initialRenderingState === RenderingStates.FINISHED
```
I might be missing something, but can we actually reach this spot if the initial RenderingState is FINISHED?

nicolo-ribaudo replied:
If the initial rengering state is FINISHED, we'll `console.error`, `.reset()` and then continue.

I don't think PDF.js itself ever calls `.draw()` on a finished page, but I assume that that check+reset is because other consumers might?

Snuffleupagus replied:
> I don't think PDF.js itself ever calls `.draw()` on a finished page, but I assume that that check+reset is because other consumers might?

Correct, that should never happen in the default viewer (so it's probably done to account for user-error when building upon e.g. the [pageviewer example](https://github.com/mozilla/pdf.js/blob/master/examples/components/pageviewer.mjs)).

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -1240,4 +1277,229 @@ class PDFPageView extends PDFPageViewBase {
+      },
+      prevCanvas,
+      () => {
+        this.eventBus.dispatch("pagerendered", {
+          source: this,
+          pageNumber: this.id,
+          cssTransform: false,
+          timestamp: performance.now(),
+          error: this._renderError,
+        });
```
Maybe we should introduce helper-methods in the base-class for dispatching this event, to reduce duplication?

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -1240,4 +1277,229 @@ class PDFPageView extends PDFPageViewBase {
+
+    div.setAttribute("data-loaded", true);
+
+    this.eventBus.dispatch("pagerender", {
+      source: this,
+      pageNumber: this.id,
+    });
```
Maybe we should introduce helper-methods in the base-class for dispatching this event, to reduce duplication?

Snuffleupagus reviewed (commented):
When we reset the "main" page and cancel its rendering here: https://github.com/mozilla/pdf.js/blob/a1c884cc469f6e0a7a78b657cbc6b7cf2dbf5413/web/pdf_page_view.js#L674-L686
don't we also need to reset and cancel a pending `detailView` rendering as well?

Snuffleupagus commented:
Btw, does this PR also fix issue #14193?

nicolo-ribaudo commented on test/integration/viewer_spec.mjs:
```diff
@@ -454,4 +454,346 @@ describe("PDF viewer", () => {
+
+      beforeAll(async () => {
+        const pagesP = loadAndWait(
+          "colors.pdf",
```
Note to myself: I forgot to upload this PDF

Snuffleupagus commented on test/integration/viewer_spec.mjs:
```diff
@@ -454,4 +454,346 @@ describe("PDF viewer", () => {
+          null,
+          zoom,
+          {
+            // TODO: Why is this needed?
+            earlySetup: () => {
+              window.devicePixelRatio = 2;
+            },
```
What sort of errors happen without this TODO?

Note that the `devicePixelRatio` value will affect the canvas-sizes in the viewer, and thus how "quickly" we reach e.g. the `maxCanvasPixels` value.
Is it possible that the initial/expected values in these tests "simply" need to take the `devicePixelRatio` into account instead?

nicolo-ribaudo replied:
If I don't explicitly set the same `devicePixelRatio` in `window.devicePixelRatio` and in the viewport options for puppeteer, then it seems like in Chrome the `window.devicePixelRatio` variable doesn't actually match the pixel ratio that the browser is running with.

My screen has a pixel ratio of 2, puppeteer defaults to 1, and they seem to somehow lead to chrome scrolling to the wrong amount of pixels when trying to scroll.

I have hard-coded 2 for now and tests are passing, but I need to figure out why there is a problem and maybe run them both with 1 and non-1.

Snuffleupagus commented on test/integration/viewer_spec.mjs:
```diff
@@ -454,4 +454,346 @@ describe("PDF viewer", () => {
+      let pages;
+
+      beforeAll(async () => {
+        const pagesP = loadAndWait(
```
Nit: Personally I'm finding these sort of one-letter abbreviations unclear when reading code, can we please just spell it out instead?
```suggestion
        const pagesPromise = loadAndWait(
```

Snuffleupagus commented on test/integration/viewer_spec.mjs:
```diff
@@ -454,4 +454,346 @@ describe("PDF viewer", () => {
+    }
+
+    function extractCanvases(pageNumber) {
+      const toHex = ([r, g, b]) =>
+        `#` +
+        r.toString(16).padStart(2, "0") +
+        g.toString(16).padStart(2, "0") +
+        b.toString(16).padStart(2, "0");
```
I suppose that it's not easily possible to use our existing helper function for this?
https://github.com/mozilla/pdf.js/blob/016de7422924e3f62d07c826965e549fcea0ff03/src/shared/util.js#L673-L675

nicolo-ribaudo replied:
No, also because this code is passed to `page.evaluate` so I cannot simply import `Util` in the file. It would require adding a `PDFJSDev.test("TESTING")` check where that `Util` class is defined, to expose it as on `globalThis.Util`.

Snuffleupagus replied:
It *should* already be available through `globalThis.pdfjsLib.Util`, but maybe that still isn't helpful here?

nicolo-ribaudo replied:
That helped :)

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -1189,4 +1397,225 @@ class PDFPageView {
+  async draw() {
+    this.recorder.start();
+
+    const initialRenderingState = this.renderingState;
+    if (initialRenderingState !== RenderingStates.INITIAL) {
+      console.error("Must be in new state before drawing");
+      this.reset(); // Ensure that we reset all state to prevent issues.
+    }
+    const { div, pdfPage, viewport } = this.pageView;
+
+    if (!pdfPage) {
+      this.renderingState = RenderingStates.FINISHED;
+      throw new Error("pdfPage is not loaded");
+    }
+
+    this.renderingState = RenderingStates.RUNNING;
+
+    const canvasWrapper = this.pageView._ensureCanvasWrapper();
+
+    const { canvas, prevCanvas, ctx } = this._createCanvas(
+      newCanvas => {
+        // If there is already the background canvas, inject this new canvas
+        // after it. We cannot simply use .append because all canvases must
+        // be before the SVG elements used for drawings.
+        if (canvasWrapper.firstElementChild?.tagName === "CANVAS") {
+          canvasWrapper.firstElementChild.after(newCanvas);
+        } else {
+          canvasWrapper.prepend(newCanvas);
+        }
+      },
+      // If there is already the lower resolution canvas behind,
+      // we don't show the new one until when it's fully ready.
+      this.pageView.renderingState === RenderingStates.FINISHED ||
+        initialRenderingState === RenderingStates.FINISHED
```
To me it seems a tiny bit easier if we'd instead just determine `hideUntilComplete` upfront here.
```suggestion
    // If there is already the lower resolution canvas behind,
    // we don't show the new one until when it's fully ready.
    const hideUntilComplete =
      this.pageView.renderingState === RenderingStates.FINISHED ||
      this.renderingState === RenderingStates.FINISHED

    if (this.renderingState !== RenderingStates.INITIAL) {
      console.error("Must be in new state before drawing");
      this.reset(); // Ensure that we reset all state to prevent issues.
    }
    const { div, pdfPage, viewport } = this.pageView;

    if (!pdfPage) {
      this.renderingState = RenderingStates.FINISHED;
      throw new Error("pdfPage is not loaded");
    }

    this.renderingState = RenderingStates.RUNNING;

    const canvasWrapper = this.pageView._ensureCanvasWrapper();

    const { canvas, prevCanvas, ctx } = this._createCanvas(
      newCanvas => {
        // If there is already the background canvas, inject this new canvas
        // after it. We cannot simply use .append because all canvases must
        // be before the SVG elements used for drawings.
        if (canvasWrapper.firstElementChild?.tagName === "CANVAS") {
          canvasWrapper.firstElementChild.after(newCanvas);
        } else {
          canvasWrapper.prepend(newCanvas);
        }
      },
      hideUntilComplete
```

nicolo-ribaudo commented:
@calixteman Updated also to use `aria-hidden="true"` in the extra canvas instead of `role="presentation"`

nicolo-ribaudo committed:
Avoid degrading scroll performance due to the detail view

When scrolling quickly, the constant re-rendering of the detail view
significantly affects rendering performance, causing Firefox to
not render even the _background canvas_, which is just a static canvas
not being re-drawn by JavaScript.

This commit changes the viewer to only render the detail view while
scrolling if its rendering hasn't just been cancelled. This means that:
- when the user is scrolling slowly, we have enough time to render the
  detail view before that we need to change its area, so the user always
  sees the full screen as high resolution.
- when the user is scrolling quickly, as soon as we have to cancel a
  rendering we just give up, and the user will see the lower resolution
  canvas. When then the user stops scrolling, we render the detail view
  for the new visible area.

nicolo-ribaudo commented:
I noticed a performance problem:

https://github.com/user-attachments/assets/a4e0bf8f-6691-4ba4-94ee-780852900bb6

When scrolling too quickly, the detail view constant re-drawing overloads Firefox, causing it to not be even able to render the background canvas that gets scrolled around. This leads to flashes of "gray", because it just shows the background behind the canvas.

The `Avoid degrading scroll performance due to the detail view` commit fixes it, by delaying rendering the detail view once we start trashing in-progress re-drawings due to the quick scrolling.

nicolo-ribaudo commented:
Q about the "release blocker" label: is it because the next version is a major, and you want to ship this in a major?

Snuffleupagus commented:
> Q about the "release blocker" label: is it because the next version is a major, and you want to ship this in a major?

It's not that we couldn't just land this in any "normal" release, but we sometimes decide to bump the major version number when there's a couple of significant new features.
Given that this PR fixes longstanding issues w.r.t. rendering quality of large pages, this definitely seems to fit that criteria along with other recent changes such as e.g. the separate JPX wasm-file.

(The label itself is merely used to avoid accidentally creating a release before everything that should go into it is done.)

nicolo-ribaudo commented:
I _think_ I fixed the Puppeteer problem. Now:
- all the tests run with 1x and 2x devicePixelRatio
- For Firefox, we still need to manually set `window.devicePixelRatio` because Puppeteer's `devicePixelRatio` does not seem to be setting it
- there was also a bug in the code, where when using `devicePixelRatio: 1` it didn't properly translate the contents of the detail canvas.

Snuffleupagus commented:
/botio integrationtest

nicolo-ribaudo commented:
This is the failure:

```
Suite error: PDF viewer Detail view on zoom setupPages()
  Message:
[31m    ProtocolError: browsingContext.activate timed out. Increase the 'protocolTimeout' setting in launch/connect calls for a higher timeout if needed.[0m
  Stack:
        at <instance_members_initializer> (file:///home/ubuntu/pdfjs/botio-files-pdfjs/private/2be5ccc19bb8013/node_modules/puppeteer-core/lib/esm/puppeteer/common/CallbackRegistry.js:97:14)
        at new Callback (file:///home/ubuntu/pdfjs/botio-files-pdfjs/private/2be5ccc19bb8013/node_modules/puppeteer-core/lib/esm/puppeteer/common/CallbackRegistry.js:101:16)
        at CallbackRegistry.create (file:///home/ubuntu/pdfjs/botio-files-pdfjs/private/2be5ccc19bb8013/node_modules/puppeteer-core/lib/esm/puppeteer/common/CallbackRegistry.js:20:26)
        at BidiConnection.send (file:///home/ubuntu/pdfjs/botio-files-pdfjs/private/2be5ccc19bb8013/node_modules/puppeteer-core/lib/esm/puppeteer/bidi/Connection.js:51:32)
        at Session.send (file:///home/ubuntu/pdfjs/botio-files-pdfjs/private/2be5ccc19bb8013/node_modules/puppeteer-core/lib/esm/puppeteer/bidi/core/Session.js:134:42)
        at Session.<anonymous> (file:///home/ubuntu/pdfjs/botio-files-pdfjs/private/2be5ccc19bb8013/node_modules/puppeteer-core/lib/esm/puppeteer/util/decorators.js:101:27)
        at BrowsingContext.activate (file:///home/ubuntu/pdfjs/botio-files-pdfjs/private/2be5ccc19bb8013/node_modules/puppeteer-core/lib/esm/puppeteer/bidi/core/BrowsingContext.js:286:33)
        at BrowsingContext.<anonymous> (file:///home/ubuntu/pdfjs/botio-files-pdfjs/private/2be5ccc19bb8013/node_modules/puppeteer-core/lib/esm/puppeteer/util/decorators.js:101:27)
        at BidiPage.bringToFront (file:///home/ubuntu/pdfjs/botio-files-pdfjs/private/2be5ccc19bb8013/node_modules/puppeteer-core/lib/esm/puppeteer/bidi/Page.js:461:47)
        at file:///home/ubuntu/pdfjs/botio-files-pdfjs/private/2be5ccc19bb8013/test/integration/test_utils.mjs:113:18
```

I had a similar failure locally, that was fixed by running the tests in the same browser for different resolutions serially rather than in parallel. It seems however that it's not enough on the bots.

Something that was even better was to wrap all the describe block in a `for (const resolution of [1, 2])`, and basically have completely separated tests for different resolutions, rather than iterating through the various resolutions _inside each test_.

Right now all the tests with multiple combinations (mostly, browsers) iterate inside the `it` blocks. Would this other approach be ok for these tests?

nicolo-ribaudo commented:
Ok lets see if that worked (I never tried using it, but I believe I now have permission for the bots commands)

nicolo-ribaudo commented:
/botio integrationtest

nicolo-ribaudo commented:
The tests for this PR now passed :tada:

There are some timeouts regarding the FreeText editor tests, and I'm getting flaky timeouts on them locally too (both on this branch, and on 8358ab63b3a331cd90e06c35f26e8907cf3f7b81, which is the commit that this branch is based on). I'll try rebasing, since I see that there are new commits that touch those tests.

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -110,24 +116,234 @@ const LAYERS_ORDER = new Map([
+      pageNumber: this.id,
+      cssTransform,
+      timestamp: performance.now(),
+      error: this._renderError,
```
Can't we just access the private field directly now?
```suggestion
      error: this.#renderError,
```

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -110,24 +116,234 @@ const LAYERS_ORDER = new Map([
+    }
+  }
+
+  get _renderError() {
+    return this.#renderError;
+  }
+
```
We the previous comment addressed, I don't believe that we need this getter any more.

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -110,24 +116,234 @@ const LAYERS_ORDER = new Map([
+    const renderTask = (this.renderTask = this.pdfPage.render(options));
+    renderTask.onContinue = this.#renderContinueCallback;
+
+    try {
+      await renderTask.promise;
+      this.#showCanvas?.(true);
+      this.#finishRenderTask(renderTask, null, onFinish);
+    } catch (error) {
+      // When zooming with a `drawingDelay` set, avoid temporarily showing
+      // a black canvas if rendering was cancelled before the `onContinue`-
+      // callback had been invoked at least once.
+      if (!(error instanceof RenderingCancelledException)) {
+        this.#showCanvas?.(true);
+      } else {
+        prevCanvas?.remove();
+        this._resetCanvas();
+      }
+      this.#finishRenderTask(renderTask, error, onFinish);
+    }
+  }
+
+  async #finishRenderTask(renderTask, error, onFinish) {
+    // The renderTask may have been replaced by a new one, so only remove
```
*Optional suggestion:* Do we still need a separate method, or can these be combined now?
```suggestion
    let error = null;
    try {
      await renderTask.promise;
      this.#showCanvas?.(true);
    } catch (ex) {
      // When zooming with a `drawingDelay` set, avoid temporarily showing
      // a black canvas if rendering was cancelled before the `onContinue`-
      // callback had been invoked at least once.
      if (!(ex instanceof RenderingCancelledException)) {
        this.#showCanvas?.(true);
      } else {
        prevCanvas?.remove();
        this._resetCanvas();
      }
      error = ex;
    }

    // The renderTask may have been replaced by a new one, so only remove
```

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -1189,4 +1299,240 @@ class PDFPageView {
+    canvas.style.left = `${(area.minX * 100) / width}%`;
+
+    const renderingPromise = this._drawCanvas(
+      {
+        canvasContext: ctx,
+        transform,
+        viewport,
+        pageColors: this.pageColors,
+      },
```
Why don't this need to include more of the parameter as in the "full page" rendering, since as-is it *appears* to me that things such as annotationMode, optionalContent, and editingState won't be correctly reflected in the detailView rendering?

https://github.com/mozilla/pdf.js/blob/79a976ce9d49c3a3510588281a386804ebdfb23a/web/pdf_page_view.js#L1070-L1079

nicolo-ribaudo replied:
Do you have an example flow of how I can test the difference?

Snuffleupagus replied:
I've just tested this in a PDF document with optionalContent, and it's unfortunately broken just like I imagined so you'll need to add *all* of those parameters to the detailView-`renderContext` as well.

As for testing e.g. the optionalContent stuff:
 - Load any PDF file with such content, e.g. `issue18823.pdf`.
 - Change, e.g. via the sidebar, which optionalContent is visible.
 - Zoom in the document to trigger detailView rendering (to see this easier I manually lowered the `maxCanvasPixels` option first).

Once the detailView renders it uses the initial optionalContent state, rather than the *current* state, which is clearly visible when zooming in/out past the detailView render-threshold.

nicolo-ribaudo commented:
I'm still working on a bug found by @calixteman about some race condition between scrolling and zooming:
![image](https://github.com/user-attachments/assets/a296f37a-4a84-4075-b17d-05ca40a0d6ab)

Sometimes, when quickly scrolling and zooming, we render the contents of the detail canvas with the right scale, but we use a previous scale to determine how large it should  be.

nicolo-ribaudo commented:
Half way. With the current release you can only see that PDF if the zoom is low, and with this PR you can also see if the zoom is high (when the detail view triggers).

I still see it as blank at "medium" zoom level, probably because the canvas is too tall for Firefox but its area is too small for the css-only/detail-view logic to apply. Probably a good fix would be to trigger the css-only/detail-view logic earlier, not just based on the canvas area but also on its width/height.

I'd leave that issue open for now.

nicolo-ribaudo commented:
The last commit _should_ fix https://github.com/mozilla/pdf.js/pull/19128#issuecomment-2605663514. Unfortunately it is very racy behavior and I'm not able to write a test.

> fixup! Render high-res partial page views when falling back to CSS zoom 
>
> Make sure that the `#detailArea` is not used from a previous computation
> with a different scale.
> 
> The `#detailArea` can become out of date when very quickly alternating
> between scrolling and zooming. This causes the canvas to be too big,
> even if then the rendered contents have the right scale.
> 
> This commit:
> 1. changes `#shouldRenderDifferentArea` so that, whenever the scale
>    changes, it forces re-computing what is the size of the new canvas
>    to render.
> 2. making sure to trash the detail area of pages that are not visible
>    anymore, so that we don't risk trying to render it before calling
>    update().

I have been scrolling and zooming around the Tracemonkey PDF for 5 minutes (with `maxCanvasPixels: 2 **20`, which makes the issue more present) and the problem didn't happen anymore, while before that commit it was happening in about 1 minute of random scrolling and zooming.

Snuffleupagus commented on web/pdf_viewer.js:
```diff
@@ -1664,6 +1682,17 @@ class PDFViewer {
     const newCacheSize = Math.max(DEFAULT_CACHE_SIZE, 2 * numVisiblePages + 1);
     this.#buffer.resize(newCacheSize, visible.ids);
 
+    const visiblePagesSet = new Set();
+    for (const { view, visibleArea } of visiblePages) {
+      view.updateVisibleArea(visibleArea);
+      visiblePagesSet.add(view);
+    }
+    for (const view of this._pages) {
+      if (!visiblePagesSet.has(view)) {
+        view.updateVisibleArea(null);
+      }
+    }
```
*Another question:* Why do we need to check *all* pages here, since having to do that seems unfortunate in long documents (e.g. with thousands of pages)?
Pages should be in one of the following states here:
 - Visible ones.
 - Invisible ones, but cached in `this.#buffer`.
 - The rest of them.

When pages are evicted from `this.#buffer` they should have all of their rendering, including the detailView, cancelled.
So unless I'm misunderstanding things, isn't it sufficient to only `null` pages that exist in `this.#buffer`?

```suggestion
    for (const { view, visibleArea } of visiblePages) {
      view.updateVisibleArea(visibleArea);
    }
    for (const view of this.#buffer) {
      if (!visible.ids(view.id)) {
        view.updateVisibleArea(null);
      }
    }
```

nicolo-ribaudo replied:
Yes this is indeed better, given the below suggestion to properly clean everything up on `.reset()`

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -583,7 +745,11 @@ class PDFPageView {
+    }
+
+    if (!preserveDetailViewState) {
+      this.detailView?.reset({ keepCanvas: keepCanvasWrapper });
```
When we're doing a "full" reset, i.e. calling `this.reset()` without any parameters (e.g. what happens when evicting pages from the `PDFPageViewBuffer`-instance), shouldn't we also set `this.detailView = null;` to clear it out?

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -583,7 +745,12 @@ class PDFPageView {
+    }
+
+    if (!preserveDetailViewState) {
+      this.detailView?.reset({ keepCanvas: keepCanvasWrapper });
```
To ensure that we leave nothing behind, don't we need to ensure that a detailView canvas also gets removed here?
```suggestion
      this.detailView?.reset();
```

nicolo-ribaudo replied:
Ok yeah the current code is wrong. When opening a PDF and switching toggling on/off an annotations editor, the detail canvas gets left behind in the DOM and never cleaned up. So then if you change the visible area, you get a _third_ canvas.

Not passing `keepCanvas: keepCanvasWrapper` is also wrong, because when toggling the annotations editor we want to keep the detail canvas around.

The correct solution is to only set `this.detailView` when _not_ `keepCanvasWrapper` is false, so that if we still have the canvas we still have the class that tracks it (pushing a commit to fix this now)

nicolo-ribaudo replied:
No, because in this case where we toggle the annotation editor we do want to cancel the in-progress rendering and update the state accordingly, but we don't want to remove the canvas until when the new one is done re-rendering.

So we _could_ pass `preserveDetailViewState: true`, but then we'd need to manually call `this.detailView?.reset({ keepCanvas: true })` after calling `this.reset()` in that place.

nicolo-ribaudo commented:
I was playing with this on Firefox on Android, and noticed that the way this PR handled the detail canvas is quite annoying: when scrolling/zooming, the rendering gets cancelled very often and thus there is no new canvas to replace the old one. The last commit fixes it. I'll attach soon a video with the before/after (before is the before the page reload, after is after the page reload).

calixteman commented on web/pdf_page_view.js:
```diff
@@ -1189,4 +1305,247 @@ class PDFPageView {
+    // finishing rendering the new detail before they get past the rendered
+    // area.
+
+    const MOVEMENT_TRESHOLD = 0.5;
```
nit: -TRESHOLD+THRESHOLD

Snuffleupagus commented on web/pdf_viewer.js:
```diff
@@ -1851,6 +1862,18 @@ class PDFViewer {
     );
 
     if (pageView) {
+      if (
+        this.#scrollTimeoutId !== null &&
+        pageView instanceof PDFPageDetailView &&
+        pageView.renderingCancelled
+      ) {
+        // If we are scrolling and the rendering of the detail view was just
+        // cancelled, it's because the user is scrolling too quickly and so
+        // we constantly need to re-render a different area.
+        // Don't attempt to re-rendering it: this will be done once the user
+        // stops scrolling.
+        return true;
+      }
```
I don't know how much it matters in practice, but won't returning here without trigger rendering effectively stop any possible pre-rendering of other *currently invisible* pages?

Until either another page becomes the more visible *or* scrolling stops, it seems that the `getHighestPriority`-method will [now always break here](https://github.com/mozilla/pdf.js/pull/19128/files#diff-06d001972b1007eb10259e1678f748a90ad984af2ab252a1125d4387d696f61cR132-R134).

Would it make (any) sense to e.g. set the detailView-renderingState to `FINISHED` when `renderingCancelled === true` such that other pre-rendering can still run?

nicolo-ribaudo replied:
I fixed this in a slightly different way, as when marking the view as finished then we don't re-render it unless the new area is enough different from the old one.

Instead, `getHighestPriority` now receives a parameter to make it ignore detail views.

Snuffleupagus commented on test/integration/viewer_spec.mjs:
```diff
@@ -454,4 +454,354 @@ describe("PDF viewer", () => {
+        window.PDFViewerApplication.eventBus.on(
+          "pagerendered",
+          ({ source }) => {
+            if (source.constructor.name === "PDFPageDetailView") {
```
Here, and elsewhere, having to rely explicitly on the constructor name feels a little "brittle" to me.

How about instead adding a `isDetailView` property to the "pagerendered" event dispatching (i.e. similar to the existing `cssTransform`)?

nicolo-ribaudo replied:
That would mean some "public API" for something that we only use in tests. Are you ok with that?

nicolo-ribaudo replied:
Actually yes, this is a good idea as consumers that are currently waiting for a page to be render will in many cases want to add an `!isDetailView` check.

Snuffleupagus replied:
I was thinking that it'd just be a boolean value (assuming that works for the tests), so we'd not expose any actual functionality that way.

Snuffleupagus replied:
> Actually yes, this is a good idea as consumers that are currently waiting for a page to be render will in many cases want to add an `!isDetailView` check.

Yes, and we may want to add a `|| evt.isDetailView` check to the early return here:
https://github.com/mozilla/pdf.js/blob/d1f62509e531b9e9f8862e3526eaa17565ff1e9c/web/pdf_viewer.js#L867-L870

Snuffleupagus commented on web/pdf_page_view_base.js:
```diff
@@ -0,0 +1,224 @@
+import { RenderingCancelledException } from "pdfjs-lib";
+import { RenderingStates } from "./ui_utils.js";
+
+export class PDFPageViewBase {
```
- Looking at our existing base-classes they all use a slightly different naming convention, so can we please be consistent here?
   ```suggestion
   class BasePDFPageView {
   ```
 - Can we please move the `export` to end of the file for consistency, since we don't really use "inline" exports in the code-base?

nicolo-ribaudo replied:
It it ok if I rename the class without renaming the file, so that the three "page view" files remain together when sorted alfabetically? (it just makes it easier to find them at glance)

Snuffleupagus replied:
To me, it does seem slightly confusing if the class- and file-names differ from each other.

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -1067,74 +991,68 @@ class PDFPageView {
+        this.#useThumbnailCanvas.regularAnnotations =
+          !renderTask.separateAnnots;
+
+        // If there is a `.detailView` that still needs to be rendered, it will
+        // dispatch the pagerendered event once it's done.
+        if (
+          !this.detailView ||
+          this.detailView.renderingState === RenderingStates.FINISHED
+        ) {
+          this.dispatchPageRendered(
+            /* cssTransform */ false,
+            /* isDetailView */ false
+          );
         }
```
When there's a detailView, won't this condition effectively prevent a "pagerendered" event from ever being dispatched with `isDetailView = false` set?

Is that behaviour correct, since it means that we may no longer have a way to detect the initial *full* page rendering and the following check could theoretically always hold (and thus leave `_onePageRenderedCapability pending)?
https://github.com/mozilla/pdf.js/pull/19128/files#diff-4238fc7e96d72f4095859850f3d24a9298ec2dc8bd96437214a02dccbd48385cR876

nicolo-ribaudo replied:
Mh true. I think when I originally wrote this PR I was trying to dispatch _one_ `pagerendered` event per page (whether it's an initial render or a re-render). With the `isDetailView` property on events it's indeed easier to just dispatch both times, and consumers can choose which ones to ignore.

nicolo-ribaudo commented:
@Snuffleupagus I added a detailed test for the `pagerendered` events, following https://github.com/mozilla/pdf.js/pull/19128#discussion_r1939321725, and that lead me to find a bug in this PR.

Before this PR, `#hasRestrictedRendering` meant "this view has already been rendered at a restircted scale and zoomed through CSS", while this PR was changing it to mean "this view will need to be scaled through CSS". The effect of this was that, when going from "small" to "very big", we'd just apply the CSS zoom to the small canvas rather than first re-rendering it at the maximum size and _then_ applying CSS zoom.

The reason the `#hasRestrictedRendering` change happened was because we need to know whether we'll need the detail view or not before that the rendering actually happens.

The last `fixup!` commit fixes this by splitting it into `#hasRestrictedRendering` (with the old meaning) and `#needsRestrictedRendering` (with the new meaning).

nicolo-ribaudo commented:
/botio test

Snuffleupagus commented on web/pdf_rendering_queue.js:
```diff
@@ -102,15 +102,22 @@ class PDFRenderingQueue {
+    views,
+    scrolledDown,
+    preRenderExtra = false,
+    ignoreDetailViews = false
```
Possibly a pointless optimization, but would it make sense to always pass `ignoreDetailViews === true` when rendering the thumbnails since those by design won't have them?

https://github.com/mozilla/pdf.js/blob/d1f62509e531b9e9f8862e3526eaa17565ff1e9c/web/pdf_thumbnail_viewer.js#L289-L296

nicolo-ribaudo replied:
I don't expect it to have a any significant effect, but doing it doesn't hurt for sure.

Snuffleupagus commented:
/botio integrationtest

nicolo-ribaudo commented:
Whops sorry, the test now should be fixed for real. It was accidentally relying on the first `expect()` call happen as soon as the PDF viewer opens, before that it finishes rendering anything. We don't actually need that `expect()`.

nicolo-ribaudo commented:
@Snuffleupagus please wait before re-running the tests — @calixteman messaged me a video where the detail canvas gets left behind even when it's not needed anymore, and I'm now investigating it.

nicolo-ribaudo commented:
Wow this was a difficult bug to track down, and to write tests for. I added two commits (that I will squash), fixing two different race conditions that causes the detail canvas to not be properly removed from the DOM when it's not needed anymore. Each commit includes a test.

The first commit fixes a race condition that can only be triggered via the API (since it requires two synchronous calls to `.updateScale`). The second commit fixes one that can also appear while manually zooming in/out quickly, and I believe it's the bug that @calixteman noticed.

I'm pasting here the message of the second `fixup!` commit, so that it doesn't get lost when squashing:
```
Clean up the PDFDetailView after cancellation synchronously.

This ensures that, when rendering is cancelled, we always restore tracking
the canvas that is still in the DOM (done in the `onCancel` callback of
`_drawCanvas`) before calling `.remove()` on whatever the tracked canvas
is.

Given this algorithm that starts with a 1000% zoom already rendered,
with a detail view canvas that we'll call #1:
- zoom 0.3x (!needs detail view)
- await 1 microtick (Promise.resolve())
- zoom 0.3x (!does not need detail view)

Before this patch, the order of operations would be:
- Create a new detail canvas #2 for the 300% zoom level
- Change detailView.canvas from #1 to #2
- await 1 microtick (between the two zoom changes)
- In .reset(), cancel the rendering
- In .reset(), call detailView.canvas.remove() (removing #2, even if the
  one in the dom is actually still #1)
- Set .detailView to `null`, because we don't need it anymore.
- await, because the `onCancel` logic was run asyncronously
- Change detailView.canvas from #2 to #1

After this patch, the order becomes:
- Create a new detail canvas #2 for the 300% zoom level
- Change detailView.canvas from #1 to #2
- await 1 microtick (between the two zoom changes)
- In .reset(), cancel the rendering
  - It runs `onCancel`, changing detailView.canvas from #2 to #1
- In .reset(), call detailView.canvas.remove() (which now removes #1)
- Set .detailView to `null`, because we don't need it anymore.
 ```

Snuffleupagus commented on src/display/api.js:
```diff
@@ -3289,17 +3291,27 @@ class PDFObjects {
+  /**
+   * A function that will be synchronously called when the rendering tasks
+   * finishes with an error (either because of an actual error, or because the
+   * rendering is cancelled)
```
```suggestion
   * rendering is cancelled).
```

Snuffleupagus commented on src/display/api.js:
```diff
@@ -1544,6 +1544,8 @@ class PDFPageProxy {
           intentState,
           reason: error instanceof Error ? error : new Error(error),
         });
+
+        internalRenderTask.task.onError?.(error);
```
To have a bit more consistency between the `onContinue` and `onError` handling, could we possibly move this into `InternalRenderTask` instead?

```diff
```diff --git a/src/display/api.js b/src/display/api.js
index 6da65d046..f5196744a 100644
--- a/src/display/api.js
+++ b/src/display/api.js
@@ -3428,13 +3428,13 @@ class InternalRenderTask {
     }
     InternalRenderTask.#canvasInUse.delete(this._canvas);

-    this.callback(
-      error ||
-        new RenderingCancelledException(
-          `Rendering cancelled, page ${this._pageIndex + 1}`,
-          extraDelay
-        )
+    error ||= new RenderingCancelledException(
+      `Rendering cancelled, page ${this._pageIndex + 1}`,
+      extraDelay
     );
+    this.callback(error);
+
+    this.task.onError?.(error);
   }

   operatorListChanged() {
```

Snuffleupagus commented:
/botio integrationtest

nicolo-ribaudo commented:
The windows-only failure for "PDF viewer CSS-only zoom triggers when going bigger than maxCanvasPixels triggers CSS-only zoom above maxCanvasPixels" _could_ be related, I'm taking a look

nicolo-ribaudo commented:
/botio integrationtest

nicolo-ribaudo commented:
Rebased to solve conflicts with the autolinking PR.

Snuffleupagus commented:
/botio integrationtest

Snuffleupagus commented:
/botio-linux preview

nicolo-ribaudo commented:
/botio integrationtest

calixteman commented on web/base_pdf_page_view.js:
```diff
@@ -0,0 +1,228 @@
+    const updateOnFirstShow = !prevCanvas && !hasHCM && !hideUntilComplete;
+
+    const canvas = document.createElement("canvas");
+    this.canvas = canvas;
```
nit: `const canvas = (this.canvas = ...`

calixteman commented on web/pdf_page_detail_view.js:
```diff
@@ -0,0 +1,273 @@
+
+    canvas.width = area.width * devicePixelRatio;
+    canvas.height = area.height * devicePixelRatio;
+    canvas.style.position = "absolute";
```
nit: maybe it can be moved in the css.

nicolo-ribaudo replied:
Oh actually it's already defined in the CSS, I can just delete these two lines.

calixteman commented on web/pdf_page_detail_view.js:
```diff
@@ -0,0 +1,273 @@
+
+    canvas.width = area.width * devicePixelRatio;
+    canvas.height = area.height * devicePixelRatio;
+    canvas.style.position = "absolute";
```
nit: `const { style } = canvas;` ...

calixteman reviewed (approved):
LGTM. Thank you.

marco-c commented:
> Once @calixteman is OK with this, I'll look through the PR one final time.

 @Snuffleupagus could you have another look now that Calixte approved the PR?

Snuffleupagus commented:
> could you have another look now that Calixte approved the PR?

I'm planning to, once the outstanding review comments are addressed.

nicolo-ribaudo commented on web/pdf_page_detail_view.js:
```diff
@@ -0,0 +1,273 @@
+    return this.pageView.pdfPage;
+  }
+
+  get renderingState() {
```
It'd be nice if it was useless, but it's needed when defining a setter. If you define a setter, the getter defaults to `get x() {}` rather than `get x() { return super.x }`.

nicolo-ribaudo commented:
All comments addressed!

Snuffleupagus reviewed (approved):
r=me, this is great stuff; thank you.

Snuffleupagus commented:
/botio integrationtest

nicolo-ribaudo commented:
Ughhhh the failure seems related, taking a look.

Snuffleupagus commented:
> Ughhhh the failure seems related, taking a look.

Maybe not, since I believe that we've *very occasionally* seen that in other PRs as well.

nicolo-ribaudo commented:
> Maybe not, since I believe that we've very occasionally seen that in other PRs as well.

Happy to get this merged and keep investigating then.

I can reproduce it locally if I comment out the `await awaitPromise(handle)` in https://github.com/mozilla/pdf.js/blob/e3ea92603d17595aad4b1c37c7113cb7dee0c90e/test/integration/viewer_spec.mjs#L353C11-L353C55, which makes me think that on the bots sometimes that promise is already resolved due to a `pagerendered` event triggered from the "zoom reset" in `beforeEach` (since there we are waiting for the first page to be done rendering, rather than all of them).

I can open another PR to verify if that's indeed the problem and fix it.

Snuffleupagus commented:
/botio-linux integrationtest
